File: ios/WNBABetPredictor/Models/Scoreboard.swift
import Foundation

struct ScoreboardTeamRef: Codable, Hashable, Sendable {
    var key: String?
    var name: String?
    var abbreviation: String?
    var record: String?
}

struct ScoreboardGame: Codable, Identifiable, Hashable, Sendable {
    var id: String
    var date: String?
    var name: String?
    var status: String?
    var statusState: String?
    var period: Int?
    var clock: String?
    var homeTeam: ScoreboardTeamRef?
    var awayTeam: ScoreboardTeamRef?
    var homeKey: String?
    var awayKey: String?
    var homeScore: Int?
    var awayScore: Int?
    var venue: String?
    var odds: MarketOdds?

    var isLive: Bool {
        statusState?.lowercased() == "in"
    }

    var isFinal: Bool {
        let state = statusState?.lowercased() ?? ""
        if state == "post" { return true }
        return status?.lowercased().contains("final") == true
    }

    var statusLabel: String {
        if let period, let clock, !clock.isEmpty, isLive {
            return "Q\(period) · \(clock)"
        }
        if let period, isLive {
            return "Q\(period)"
        }
        if let clock, isLive, !clock.isEmpty {
            return clock
        }
        return status ?? "Scheduled"
    }
}

enum ESPNOddsParser {
    static func parse(from competition: [String: Any]) -> MarketOdds? {
        guard let rawOdds = (competition["odds"] as? [[String: Any]])?.first else { return nil }

        let pointSpread = rawOdds["pointSpread"] as? [String: Any]
        let total = rawOdds["total"] as? [String: Any]
        let moneyline = rawOdds["moneyline"] as? [String: Any]

        let homeFavorite = ((rawOdds["homeTeamOdds"] as? [String: Any])?["favorite"] as? Bool) == true
        let awayFavorite = ((rawOdds["awayTeamOdds"] as? [String: Any])?["favorite"] as? Bool) == true
        let rawSpread = number(from: rawOdds["spread"])
        let derivedHomeSpread = rawSpread.map { spread -> Double in
            if homeFavorite { return -abs(spread) }
            if awayFavorite { return abs(spread) }
            return spread
        }

        let closeSpread = number(from: ((pointSpread?["home"] as? [String: Any])?["close"] as? [String: Any])?["line"])
            ?? derivedHomeSpread
        let openSpread = number(from: ((pointSpread?["home"] as? [String: Any])?["open"] as? [String: Any])?["line"])

        let closeTotal = number(from: ((total?["over"] as? [String: Any])?["close"] as? [String: Any])?["line"])
            ?? number(from: rawOdds["overUnder"])
        let openTotal = number(from: ((total?["over"] as? [String: Any])?["open"] as? [String: Any])?["line"])

        let homeMoneyline = americanOdds(from: ((moneyline?["home"] as? [String: Any])?["close"] as? [String: Any])?["odds"])
        let awayMoneyline = americanOdds(from: ((moneyline?["away"] as? [String: Any])?["close"] as? [String: Any])?["odds"])
        let openHomeMoneyline = americanOdds(from: ((moneyline?["home"] as? [String: Any])?["open"] as? [String: Any])?["odds"])
        let openAwayMoneyline = americanOdds(from: ((moneyline?["away"] as? [String: Any])?["open"] as? [String: Any])?["odds"])

        let hasAnyLine = closeSpread != nil || closeTotal != nil || homeMoneyline != nil || awayMoneyline != nil
        guard hasAnyLine else { return nil }

        return MarketOdds(
            provider: ((rawOdds["provider"] as? [String: Any])?["displayName"] as? String)
                ?? ((rawOdds["provider"] as? [String: Any])?["name"] as? String),
            spread: closeSpread,
            total: closeTotal,
            homeMoneyline: homeMoneyline,
            awayMoneyline: awayMoneyline,
            openingSpread: openSpread,
            openingTotal: openTotal,
            openingHomeMoneyline: openHomeMoneyline,
            openingAwayMoneyline: openAwayMoneyline,
            source: "espn",
            deepLink: (rawOdds["link"] as? [String: Any])?["href"] as? String
        )
    }

    private static func americanOdds(from raw: Any?) -> Int? {
        guard let string = raw as? String else {
            if let intValue = raw as? Int { return intValue }
            return nil
        }
        return Int(string.trimmingCharacters(in: .whitespacesAndNewlines))
    }

    private static func number(from raw: Any?) -> Double? {
        if let value = raw as? Double { return value }
        if let value = raw as? Int { return Double(value) }
        guard let string = raw as? String else { return nil }
        let filtered = string.filter { "-+0123456789.".contains($0) }
        return Double(filtered)
    }
}

struct TeamSeasonStats: Codable, Hashable, Sendable {
    var ppg: Double?
    var oppPpg: Double?
    var fgPct: Double?
    var fg3Pct: Double?
    var ftPct: Double?
    var rebounds: Double?
    var assists: Double?
    var turnovers: Double?
    var netRating: Double?
    var pace: Double?
    var gamesPlayed: Int?
}

struct PlayerSeasonStats: Codable, Identifiable, Hashable, Sendable {
    var id: String
    var name: String
    var position: String?
    var jersey: String?
    var mpg: Double?
    var ppg: Double?
    var rpg: Double?
    var apg: Double?
    var fgPct: Double?
}

struct ScoreboardResponse: Codable, Sendable {
    var games: [ScoreboardGame]
    var source: String?
    var lastUpdated: String?
    var isLive: Bool?
    var warning: String?
    var error: String?

    enum CodingKeys: String, CodingKey {
        case games, events, source, lastUpdated, isLive, warning, error
    }

    init(
        games: [ScoreboardGame],
        source: String? = nil,
        lastUpdated: String? = nil,
        isLive: Bool? = nil,
        warning: String? = nil,
        error: String? = nil
    ) {
        self.games = games
        self.source = source
        self.lastUpdated = lastUpdated
        self.isLive = isLive
        self.warning = warning
        self.error = error
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        if let games = try container.decodeIfPresent([ScoreboardGame].self, forKey: .games) {
            self.games = games
        } else if let events = try container.decodeIfPresent([ScoreboardGame].self, forKey: .events) {
            self.games = events
        } else {
            self.games = []
        }
        source = try container.decodeIfPresent(String.self, forKey: .source)
        lastUpdated = try container.decodeIfPresent(String.self, forKey: .lastUpdated)
        isLive = try container.decodeIfPresent(Bool.self, forKey: .isLive)
        warning = try container.decodeIfPresent(String.self, forKey: .warning)
        error = try container.decodeIfPresent(String.self, forKey: .error)
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(games, forKey: .games)
        try container.encodeIfPresent(source, forKey: .source)
        try container.encodeIfPresent(lastUpdated, forKey: .lastUpdated)
        try container.encodeIfPresent(isLive, forKey: .isLive)
        try container.encodeIfPresent(warning, forKey: .warning)
        try container.encodeIfPresent(error, forKey: .error)
    }
}

struct TeamStatsDetailPayload: Codable, Sendable {
    var teamKey: String
    var teamName: String?
    var record: String?
    var abbreviation: String?
    var stats: TeamSeasonStats?
    var team: ScoreboardTeamRef?
    var source: String?
    var lastUpdated: String?
    var isLive: Bool?
    var warning: String?
    var error: String?
}

struct TeamPlayersPayload: Codable, Sendable {
    var teamKey: String
    var players: [PlayerSeasonStats]
    var source: String?
    var lastUpdated: String?
    var isLive: Bool?
    var warning: String?

    enum CodingKeys: String, CodingKey {
        case teamKey, players, roster, source, lastUpdated, isLive, warning
    }

    init(
        teamKey: String,
        players: [PlayerSeasonStats],
        source: String? = nil,
        lastUpdated: String? = nil,
        isLive: Bool? = nil,
        warning: String? = nil
    ) {
        self.teamKey = teamKey
        self.players = players
        self.source = source
        self.lastUpdated = lastUpdated
        self.isLive = isLive
        self.warning = warning
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        teamKey = try container.decode(String.self, forKey: .teamKey)
        if let players = try container.decodeIfPresent([PlayerSeasonStats].self, forKey: .players) {
            self.players = players
        } else {
            self.players = try container.decodeIfPresent([PlayerSeasonStats].self, forKey: .roster) ?? []
        }
        source = try container.decodeIfPresent(String.self, forKey: .source)
        lastUpdated = try container.decodeIfPresent(String.self, forKey: .lastUpdated)
        isLive = try container.decodeIfPresent(Bool.self, forKey: .isLive)
        warning = try container.decodeIfPresent(String.self, forKey: .warning)
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(teamKey, forKey: .teamKey)
        try container.encode(players, forKey: .players)
        try container.encodeIfPresent(source, forKey: .source)
        try container.encodeIfPresent(lastUpdated, forKey: .lastUpdated)
        try container.encodeIfPresent(isLive, forKey: .isLive)
        try container.encodeIfPresent(warning, forKey: .warning)
    }
}

File: server/scoreboard.js
'use strict';

const cache = require('./cache');
const dataFetcher = require('./data-fetcher');
const espn = require('./espn');

const ESPN_BASE = 'https://site.api.espn.com/apis/site/v2/sports/basketball/wnba';

function todayDateStr() {
  return new Date().toISOString().slice(0, 10);
}

function attachMeta(payload, cached, overrides = {}) {
  return {
    ...payload,
    source: overrides.source ?? payload.source,
    lastUpdated: overrides.lastUpdated ?? payload.lastUpdated ?? new Date().toISOString(),
    cacheAgeSeconds: cached?.cacheAgeSeconds ?? 0,
    isLive: overrides.isLive ?? payload.isLive ?? false,
    warning: overrides.warning ?? payload.warning ?? null,
  };
}

function normalizeCompetitor(comp) {
  if (!comp) return null;
  const team = comp.team ? espn.teamKeyFromAbbrev(comp.team.abbreviation) : null;
  return {
    key: team,
    name: comp.team?.displayName || comp.team?.name || null,
    abbreviation: comp.team?.abbreviation || null,
    score: parseInt(comp.score, 10) || 0,
    record: comp.records?.[0]?.summary || comp.record || null,
    winner: comp.winner === true,
  };
}

function numberFromLine(value) {
  if (typeof value === 'number') return value;
  if (typeof value === 'string') {
    const filtered = value.replace(/[^0-9+-.]/g, '');
    const parsed = parseFloat(filtered);
    return Number.isFinite(parsed) ? parsed : null;
  }
  return null;
}

function americanFromOdds(value) {
  if (typeof value === 'number') return value;
  if (typeof value === 'string') {
    const parsed = parseInt(value, 10);
    return Number.isFinite(parsed) ? parsed : null;
  }
  return null;
}

function normalizeOdds(competition) {
  const raw = competition?.odds?.[0];
  if (!raw) return null;

  const pointSpread = raw.pointSpread || {};
  const total = raw.total || {};
  const moneyline = raw.moneyline || {};
  const homeFavorite = raw.homeTeamOdds?.favorite === true;
  const awayFavorite = raw.awayTeamOdds?.favorite === true;
  const baseSpread = numberFromLine(raw.spread);
  const homeSpread =
    numberFromLine(pointSpread.home?.close?.line) ??
    (baseSpread != null ? (homeFavorite ? -Math.abs(baseSpread) : awayFavorite ? Math.abs(baseSpread) : baseSpread) : null);

  const odds = {
    provider: raw.provider?.displayName || raw.provider?.name || null,
    spread: homeSpread,
    total: numberFromLine(total.over?.close?.line) ?? numberFromLine(raw.overUnder),
    homeMoneyline: americanFromOdds(moneyline.home?.close?.odds),
    awayMoneyline: americanFromOdds(moneyline.away?.close?.odds),
    openingSpread: numberFromLine(pointSpread.home?.open?.line),
    openingTotal: numberFromLine(total.over?.open?.line),
    openingHomeMoneyline: americanFromOdds(moneyline.home?.open?.odds),
    openingAwayMoneyline: americanFromOdds(moneyline.away?.open?.odds),
    source: 'espn',
    deepLink: raw.link?.href || null,
  };

  const hasLine = odds.spread != null || odds.total != null || odds.homeMoneyline != null || odds.awayMoneyline != null;
  return hasLine ? odds : null;
}

function normalizeGame(event) {
  const comp = event.competitions?.[0];
  const statusObj = comp?.status || event.status || {};
  const state = statusObj.type?.state || 'pre';
  const period = statusObj.period ?? null;

  const home = comp?.competitors?.find((c) => c.homeAway === 'home');
  const away = comp?.competitors?.find((c) => c.homeAway === 'away');

  return {
    id: event.id,
    date: event.date,
    name: event.name,
    status: statusObj.type?.description || 'Scheduled',
    statusState: state,
    statusDetail: statusObj.type?.detail || statusObj.type?.shortDetail || null,
    shortStatus: statusObj.type?.shortDetail || null,
    period,
    quarter: period,
    clock: statusObj.displayClock || null,
    clockSeconds: statusObj.clock ?? null,
    isLive: state === 'in',
    completed: statusObj.type?.completed === true || state === 'post',
    homeTeam: normalizeCompetitor(home),
    awayTeam: normalizeCompetitor(away),
    homeScore: home ? parseInt(home.score, 10) || 0 : 0,
    awayScore: away ? parseInt(away.score, 10) || 0 : 0,
    venue: comp?.venue?.fullName || null,
    broadcast: comp?.broadcasts?.[0]?.names?.join(', ') || null,
    odds: normalizeOdds(comp),
  };
}

async function fetchEspnScoreboard(dateStr) {
  const compact = dateStr.replace(/-/g, '');
  const res = await fetch(`${ESPN_BASE}/scoreboard?dates=${compact}`, {
    headers: { Accept: 'application/json', 'User-Agent': 'WNBA-Bet-Predictor/1.5' },
  });
  if (!res.ok) throw new Error(`ESPN scoreboard ${res.status}`);
  const data = await res.json();
  const games = (data.events || []).map(normalizeGame);
  return {
    date: dateStr,
    games,
    count: games.length,
    liveCount: games.filter((g) => g.isLive).length,
    source: 'espn',
    lastUpdated: new Date().toISOString(),
    isLive: true,
    warning: games.length === 0 ? `No games on ESPN scoreboard for ${dateStr}` : null,
  };
}

function scoreboardFromFallback(dateStr) {
  const fallback = dataFetcher.readFallback('schedule-fallback.json');
  const target = dateStr;
  const games = (fallback.events || [])
    .filter((e) => new Date(e.date).toISOString().slice(0, 10) === target)
    .map((e) => ({
      id: e.id,
      date: e.date,
      name: e.name,
      status: e.status || 'Scheduled',
      statusState: e.statusState || 'pre',
      statusDetail: null,
      shortStatus: e.status || null,
      period: null,
      quarter: null,
      clock: null,
      clockSeconds: null,
      isLive: e.statusState === 'in',
      completed: e.statusState === 'post',
      homeTeam: e.homeTeam
        ? { key: e.homeTeam.key, name: e.homeTeam.name, abbreviation: null, score: e.homeScore || 0 }
        : null,
      awayTeam: e.awayTeam
        ? { key: e.awayTeam.key, name: e.awayTeam.name, abbreviation: null, score: e.awayScore || 0 }
        : null,
      homeScore: e.homeScore || 0,
      awayScore: e.awayScore || 0,
      venue: e.venue || null,
      broadcast: null,
      odds: e.odds || null,
    }));

  return {
    date: dateStr,
    games,
    count: games.length,
    liveCount: games.filter((g) => g.isLive).length,
    source: fallback.source || 'local-fallback',
    lastUpdated: fallback.lastUpdated || new Date().toISOString(),
    isLive: false,
    warning: 'Local schedule fallback — live clock/scores unavailable',
  };
}

async function getScoreboardForDate(dateStr) {
  const cacheKey = dateStr;
  const cached = cache.get('scoreboard', cacheKey);
  if (cached) return attachMeta(cached.value, cached);

  if (!cache.canRefresh('scoreboard', cacheKey)) {
    const stale = cache.get('scoreboard', `${cacheKey}:stale`);
    if (stale) {
      return attachMeta(stale.value, stale, {
        warning: 'Refresh throttled — serving stale scoreboard',
        isLive: false,
      });
    }
  }

  cache.markRefresh('scoreboard', cacheKey);
  try {
    const result = await fetchEspnScoreboard(dateStr);
    cache.set('scoreboard', cacheKey, result);
    cache.set('scoreboard', `${cacheKey}:stale`, result);
    return attachMeta(result, { cacheAgeSeconds: 0 });
  } catch (err) {
    const fallback = scoreboardFromFallback(dateStr);
    fallback.warning = `ESPN scoreboard failed (${err.message}). ${fallback.warning}`;
    cache.set('scoreboard', cacheKey, fallback);
    return attachMeta(fallback, { cacheAgeSeconds: 0 }, { isLive: false });
  }
}

async function getScoreboard(dateStr = todayDateStr()) {
  const result = await getScoreboardForDate(dateStr);
  return {
    ...result,
    meta: {
      source: result.source,
      lastUpdated: result.lastUpdated,
      cacheAgeSeconds: result.cacheAgeSeconds,
      isLive: result.isLive,
      warning: result.warning,
    },
  };
}

async function getLiveScoreboard(dateStr = todayDateStr()) {
  const board = await getScoreboard(dateStr);
  const games = (board.games || []).filter((g) => g.isLive || g.statusState === 'in');
  return {
    date: board.date,
    games,
    count: games.length,
    source: board.source,
    lastUpdated: board.lastUpdated,
    cacheAgeSeconds: board.cacheAgeSeconds,
    isLive: board.isLive,
    warning: games.length === 0 ? 'No in-progress games right now' : board.warning,
    meta: {
      source: board.source,
      lastUpdated: board.lastUpdated,
      cacheAgeSeconds: board.cacheAgeSeconds,
      isLive: board.isLive,
      warning: games.length === 0 ? 'No in-progress games right now' : board.warning,
    },
  };
}

module.exports = {
  getScoreboard,
  getLiveScoreboard,
  normalizeGame,
  todayDateStr,
};
